{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## AD ML binary classification\n",
    "\n",
    "\"Supervised Machine Learning (ML) algorithms were used to identify the discriminative power of the top ten differentially expressed genes in distinguishing between the two categories. Support vector machines using both linear and radial basis function kernels were selected for the binary classification in addition to random forest and quadratic Bayes algorithms [36]. The data was divided into a training set and a test set. The training set consisted of approximately 70% of the samples (114 samples) and the remaining 30% were used for testing to test and approximate how the classifiers generalize to unknown data. There was no overlap in the subjects and samples between the training and testing data to avoid any correlation between samples in both sets. The genes were used as features for each sample with all different combinations of N genes (2 ≤ N ≤ 10) out of 10 genes. For every combination of N genes, the combination with the maximum accuracy on the training data was chosen and the corresponding accuracy on the test data was reported on those N genes to avoid data snooping. Precision and recall scores in addition to the F1 score were reported for the testing accuracy that corresponded to the maximum training accuracy over all combinations.\"\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7717689/\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This version keeps **ALL** patients and samples and does **NOT** exclude patients with TBI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pickle as pkl\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "from sklearn.dummy import DummyClassifier\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.ensemble import GradientBoostingClassifier\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.metrics import average_precision_score\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.feature_selection import RFE\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Prep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = pd.read_csv('../data/raw/gene_expression_matrix_2016-03-03/columns-samples.csv')\n",
    "donor_info = pd.read_csv('../data/raw/DonorInformation.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rnaseq_profile_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>donor_name</th>\n",
       "      <th>specimen_id</th>\n",
       "      <th>specimen_name</th>\n",
       "      <th>rna_well_id</th>\n",
       "      <th>polygon_id</th>\n",
       "      <th>structure_id</th>\n",
       "      <th>structure_acronym</th>\n",
       "      <th>structure_color</th>\n",
       "      <th>structure_name</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>Condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>488395315</td>\n",
       "      <td>309335467</td>\n",
       "      <td>H14.09.030</td>\n",
       "      <td>309357843</td>\n",
       "      <td>H14.09.030.TCx.01</td>\n",
       "      <td>395325172</td>\n",
       "      <td>320817998</td>\n",
       "      <td>10235</td>\n",
       "      <td>TCx</td>\n",
       "      <td>#ebbfd0</td>\n",
       "      <td>temporal neocortex</td>\n",
       "      <td>left</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>496100277</td>\n",
       "      <td>309335441</td>\n",
       "      <td>H14.09.004</td>\n",
       "      <td>309357624</td>\n",
       "      <td>H14.09.004.PCx.01</td>\n",
       "      <td>320630866</td>\n",
       "      <td>310967169</td>\n",
       "      <td>10557</td>\n",
       "      <td>FWM</td>\n",
       "      <td>#f2f1f0</td>\n",
       "      <td>white matter of forebrain</td>\n",
       "      <td>right</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>496100278</td>\n",
       "      <td>309335438</td>\n",
       "      <td>H14.09.001</td>\n",
       "      <td>309357596</td>\n",
       "      <td>H14.09.001.PCx.01</td>\n",
       "      <td>320630834</td>\n",
       "      <td>310790571</td>\n",
       "      <td>10557</td>\n",
       "      <td>FWM</td>\n",
       "      <td>#f2f1f0</td>\n",
       "      <td>white matter of forebrain</td>\n",
       "      <td>left</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>496100279</td>\n",
       "      <td>309335438</td>\n",
       "      <td>H14.09.001</td>\n",
       "      <td>309357599</td>\n",
       "      <td>H14.09.001.TCx.01</td>\n",
       "      <td>320630838</td>\n",
       "      <td>310790522</td>\n",
       "      <td>10235</td>\n",
       "      <td>TCx</td>\n",
       "      <td>#ebbfd0</td>\n",
       "      <td>temporal neocortex</td>\n",
       "      <td>left</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>496100281</td>\n",
       "      <td>309335439</td>\n",
       "      <td>H14.09.002</td>\n",
       "      <td>309357603</td>\n",
       "      <td>H14.09.002.HIP.01</td>\n",
       "      <td>320630842</td>\n",
       "      <td>310790372</td>\n",
       "      <td>10294</td>\n",
       "      <td>HIP</td>\n",
       "      <td>#bfb5d5</td>\n",
       "      <td>hippocampus (hippocampal formation)</td>\n",
       "      <td>right</td>\n",
       "      <td>dementia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>496100667</td>\n",
       "      <td>467056391</td>\n",
       "      <td>H15.09.103</td>\n",
       "      <td>467179071</td>\n",
       "      <td>H15.09.103.TCx.01</td>\n",
       "      <td>482655826</td>\n",
       "      <td>480366830</td>\n",
       "      <td>10235</td>\n",
       "      <td>TCx</td>\n",
       "      <td>#ebbfd0</td>\n",
       "      <td>temporal neocortex</td>\n",
       "      <td>right</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>496100669</td>\n",
       "      <td>467056391</td>\n",
       "      <td>H15.09.103</td>\n",
       "      <td>467179068</td>\n",
       "      <td>H15.09.103.PCx.01</td>\n",
       "      <td>482655822</td>\n",
       "      <td>480363830</td>\n",
       "      <td>10557</td>\n",
       "      <td>FWM</td>\n",
       "      <td>#f2f1f0</td>\n",
       "      <td>white matter of forebrain</td>\n",
       "      <td>right</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>496100670</td>\n",
       "      <td>467056406</td>\n",
       "      <td>H15.09.107</td>\n",
       "      <td>467179104</td>\n",
       "      <td>H15.09.107.TCx.01</td>\n",
       "      <td>482655780</td>\n",
       "      <td>480363840</td>\n",
       "      <td>10235</td>\n",
       "      <td>TCx</td>\n",
       "      <td>#ebbfd0</td>\n",
       "      <td>temporal neocortex</td>\n",
       "      <td>right</td>\n",
       "      <td>dementia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>496100671</td>\n",
       "      <td>467056391</td>\n",
       "      <td>H15.09.103</td>\n",
       "      <td>467179065</td>\n",
       "      <td>H15.09.103.HIP.01</td>\n",
       "      <td>482655820</td>\n",
       "      <td>480366825</td>\n",
       "      <td>10294</td>\n",
       "      <td>HIP</td>\n",
       "      <td>#bfb5d5</td>\n",
       "      <td>hippocampus (hippocampal formation)</td>\n",
       "      <td>right</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>496100672</td>\n",
       "      <td>467056407</td>\n",
       "      <td>H15.09.108</td>\n",
       "      <td>467179110</td>\n",
       "      <td>H15.09.108.PCx.01</td>\n",
       "      <td>482655784</td>\n",
       "      <td>480363849</td>\n",
       "      <td>10557</td>\n",
       "      <td>FWM</td>\n",
       "      <td>#f2f1f0</td>\n",
       "      <td>white matter of forebrain</td>\n",
       "      <td>left</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>377 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     rnaseq_profile_id   donor_id  donor_name  specimen_id      specimen_name   \n",
       "0            488395315  309335467  H14.09.030    309357843  H14.09.030.TCx.01  \\\n",
       "1            496100277  309335441  H14.09.004    309357624  H14.09.004.PCx.01   \n",
       "2            496100278  309335438  H14.09.001    309357596  H14.09.001.PCx.01   \n",
       "3            496100279  309335438  H14.09.001    309357599  H14.09.001.TCx.01   \n",
       "4            496100281  309335439  H14.09.002    309357603  H14.09.002.HIP.01   \n",
       "..                 ...        ...         ...          ...                ...   \n",
       "372          496100667  467056391  H15.09.103    467179071  H15.09.103.TCx.01   \n",
       "373          496100669  467056391  H15.09.103    467179068  H15.09.103.PCx.01   \n",
       "374          496100670  467056406  H15.09.107    467179104  H15.09.107.TCx.01   \n",
       "375          496100671  467056391  H15.09.103    467179065  H15.09.103.HIP.01   \n",
       "376          496100672  467056407  H15.09.108    467179110  H15.09.108.PCx.01   \n",
       "\n",
       "     rna_well_id  polygon_id  structure_id structure_acronym structure_color   \n",
       "0      395325172   320817998         10235               TCx         #ebbfd0  \\\n",
       "1      320630866   310967169         10557               FWM         #f2f1f0   \n",
       "2      320630834   310790571         10557               FWM         #f2f1f0   \n",
       "3      320630838   310790522         10235               TCx         #ebbfd0   \n",
       "4      320630842   310790372         10294               HIP         #bfb5d5   \n",
       "..           ...         ...           ...               ...             ...   \n",
       "372    482655826   480366830         10235               TCx         #ebbfd0   \n",
       "373    482655822   480363830         10557               FWM         #f2f1f0   \n",
       "374    482655780   480363840         10235               TCx         #ebbfd0   \n",
       "375    482655820   480366825         10294               HIP         #bfb5d5   \n",
       "376    482655784   480363849         10557               FWM         #f2f1f0   \n",
       "\n",
       "                          structure_name hemisphere Condition  \n",
       "0                     temporal neocortex       left   control  \n",
       "1              white matter of forebrain      right   control  \n",
       "2              white matter of forebrain       left   control  \n",
       "3                     temporal neocortex       left   control  \n",
       "4    hippocampus (hippocampal formation)      right  dementia  \n",
       "..                                   ...        ...       ...  \n",
       "372                   temporal neocortex      right   control  \n",
       "373            white matter of forebrain      right   control  \n",
       "374                   temporal neocortex      right  dementia  \n",
       "375  hippocampus (hippocampal formation)      right   control  \n",
       "376            white matter of forebrain       left   control  \n",
       "\n",
       "[377 rows x 13 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Process donor info to segregate control group\n",
    "control_group_df = donor_info[donor_info['act_demented'] == 'No Dementia']\n",
    "dementia_group_df = donor_info[donor_info['act_demented'] != 'No Dementia']\n",
    "\n",
    "# Get donor ids\n",
    "control_ids = control_group_df['donor_id']\n",
    "dementia_ids = dementia_group_df['donor_id']\n",
    "\n",
    "# Assign condition to sample data\n",
    "samples['Condition'] = samples['donor_id'].apply(lambda x: 'control' if x in control_ids.values else 'dementia')\n",
    "samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Grab unique donor ids for proper data split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_ids = samples.donor_id.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 70-15-15 Train-Validate-Test donor_id split\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(75, 30, 27)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 70, 15, 15 Train, Validate, Test split\n",
    "np.random.seed(42)\n",
    "train_ids = np.random.choice(donor_ids, int(np.ceil(len(donor_ids) * 0.7)))\n",
    "test_ids = np.setdiff1d(donor_ids, train_ids)\n",
    "validate_ids = np.random.choice(test_ids, int(np.ceil(len(test_ids) * 0.5)))\n",
    "test_ids = np.setdiff1d(test_ids, validate_ids)\n",
    "len(train_ids), len(test_ids), len(validate_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check condition distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Condition\n",
       "control     107\n",
       "dementia     83\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples[samples['donor_id'].isin(train_ids)]['Condition'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Condition\n",
       "dementia    48\n",
       "control     32\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples[samples['donor_id'].isin(validate_ids)]['Condition'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Condition\n",
       "control     58\n",
       "dementia    49\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples[samples['donor_id'].isin(test_ids)]['Condition'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create ML dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = pd.read_pickle('../data/interim/PyDeseq2/ct_matrix.pkl')\n",
    "top_genes = pd.read_pickle('../data/interim/PyDeseq2/top_bottom_ten_sigs.pkl')\n",
    "lv_genes = pd.read_pickle('../data/interim/lv_genes')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before dropping duplicates: 40\n",
      "After dropping duplicates: 32\n"
     ]
    }
   ],
   "source": [
    "print('Before dropping duplicates:',len(pd.concat((top_genes.gene_id, lv_genes.iloc[:,0]))))\n",
    "print('After dropping duplicates:', len(pd.concat((top_genes.gene_id, lv_genes.iloc[:,0])).drop_duplicates()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = pd.concat((top_genes.gene_id, lv_genes.iloc[:,0])).drop_duplicates().values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>rnaseq_profile_id</th>\n",
       "      <th>488395315</th>\n",
       "      <th>496100277</th>\n",
       "      <th>496100278</th>\n",
       "      <th>496100279</th>\n",
       "      <th>496100281</th>\n",
       "      <th>496100283</th>\n",
       "      <th>496100284</th>\n",
       "      <th>496100285</th>\n",
       "      <th>496100287</th>\n",
       "      <th>496100288</th>\n",
       "      <th>...</th>\n",
       "      <th>496100661</th>\n",
       "      <th>496100663</th>\n",
       "      <th>496100664</th>\n",
       "      <th>496100665</th>\n",
       "      <th>496100666</th>\n",
       "      <th>496100667</th>\n",
       "      <th>496100669</th>\n",
       "      <th>496100670</th>\n",
       "      <th>496100671</th>\n",
       "      <th>496100672</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id_mapped</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>499309627</th>\n",
       "      <td>21</td>\n",
       "      <td>85</td>\n",
       "      <td>83</td>\n",
       "      <td>44</td>\n",
       "      <td>40</td>\n",
       "      <td>34</td>\n",
       "      <td>39</td>\n",
       "      <td>32</td>\n",
       "      <td>26</td>\n",
       "      <td>27</td>\n",
       "      <td>...</td>\n",
       "      <td>43</td>\n",
       "      <td>45</td>\n",
       "      <td>24</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>53</td>\n",
       "      <td>52</td>\n",
       "      <td>26</td>\n",
       "      <td>56</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499315162</th>\n",
       "      <td>51</td>\n",
       "      <td>143</td>\n",
       "      <td>275</td>\n",
       "      <td>70</td>\n",
       "      <td>112</td>\n",
       "      <td>80</td>\n",
       "      <td>57</td>\n",
       "      <td>45</td>\n",
       "      <td>28</td>\n",
       "      <td>100</td>\n",
       "      <td>...</td>\n",
       "      <td>76</td>\n",
       "      <td>95</td>\n",
       "      <td>151</td>\n",
       "      <td>28</td>\n",
       "      <td>105</td>\n",
       "      <td>87</td>\n",
       "      <td>78</td>\n",
       "      <td>66</td>\n",
       "      <td>117</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499315537</th>\n",
       "      <td>13</td>\n",
       "      <td>22</td>\n",
       "      <td>14</td>\n",
       "      <td>49</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>48</td>\n",
       "      <td>44</td>\n",
       "      <td>14</td>\n",
       "      <td>48</td>\n",
       "      <td>...</td>\n",
       "      <td>70</td>\n",
       "      <td>43</td>\n",
       "      <td>51</td>\n",
       "      <td>9</td>\n",
       "      <td>24</td>\n",
       "      <td>103</td>\n",
       "      <td>103</td>\n",
       "      <td>11</td>\n",
       "      <td>124</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499315843</th>\n",
       "      <td>18</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>25</td>\n",
       "      <td>18</td>\n",
       "      <td>15</td>\n",
       "      <td>20</td>\n",
       "      <td>19</td>\n",
       "      <td>...</td>\n",
       "      <td>26</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>39</td>\n",
       "      <td>20</td>\n",
       "      <td>23</td>\n",
       "      <td>29</td>\n",
       "      <td>35</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499317231</th>\n",
       "      <td>31</td>\n",
       "      <td>15</td>\n",
       "      <td>30</td>\n",
       "      <td>42</td>\n",
       "      <td>55</td>\n",
       "      <td>38</td>\n",
       "      <td>35</td>\n",
       "      <td>62</td>\n",
       "      <td>37</td>\n",
       "      <td>39</td>\n",
       "      <td>...</td>\n",
       "      <td>64</td>\n",
       "      <td>89</td>\n",
       "      <td>47</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>73</td>\n",
       "      <td>42</td>\n",
       "      <td>48</td>\n",
       "      <td>46</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499317260</th>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499319747</th>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>16</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>39</td>\n",
       "      <td>24</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>39</td>\n",
       "      <td>26</td>\n",
       "      <td>6</td>\n",
       "      <td>21</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499322427</th>\n",
       "      <td>638</td>\n",
       "      <td>216</td>\n",
       "      <td>184</td>\n",
       "      <td>806</td>\n",
       "      <td>435</td>\n",
       "      <td>222</td>\n",
       "      <td>287</td>\n",
       "      <td>713</td>\n",
       "      <td>491</td>\n",
       "      <td>513</td>\n",
       "      <td>...</td>\n",
       "      <td>702</td>\n",
       "      <td>1357</td>\n",
       "      <td>1418</td>\n",
       "      <td>308</td>\n",
       "      <td>976</td>\n",
       "      <td>1059</td>\n",
       "      <td>681</td>\n",
       "      <td>512</td>\n",
       "      <td>648</td>\n",
       "      <td>742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499323096</th>\n",
       "      <td>102</td>\n",
       "      <td>235</td>\n",
       "      <td>205</td>\n",
       "      <td>136</td>\n",
       "      <td>244</td>\n",
       "      <td>204</td>\n",
       "      <td>191</td>\n",
       "      <td>132</td>\n",
       "      <td>157</td>\n",
       "      <td>192</td>\n",
       "      <td>...</td>\n",
       "      <td>117</td>\n",
       "      <td>106</td>\n",
       "      <td>197</td>\n",
       "      <td>261</td>\n",
       "      <td>225</td>\n",
       "      <td>129</td>\n",
       "      <td>146</td>\n",
       "      <td>250</td>\n",
       "      <td>169</td>\n",
       "      <td>183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499323166</th>\n",
       "      <td>462</td>\n",
       "      <td>362</td>\n",
       "      <td>297</td>\n",
       "      <td>466</td>\n",
       "      <td>545</td>\n",
       "      <td>725</td>\n",
       "      <td>580</td>\n",
       "      <td>373</td>\n",
       "      <td>586</td>\n",
       "      <td>904</td>\n",
       "      <td>...</td>\n",
       "      <td>464</td>\n",
       "      <td>372</td>\n",
       "      <td>519</td>\n",
       "      <td>938</td>\n",
       "      <td>606</td>\n",
       "      <td>526</td>\n",
       "      <td>508</td>\n",
       "      <td>859</td>\n",
       "      <td>620</td>\n",
       "      <td>547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499324505</th>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>108</td>\n",
       "      <td>43</td>\n",
       "      <td>68</td>\n",
       "      <td>59</td>\n",
       "      <td>119</td>\n",
       "      <td>45</td>\n",
       "      <td>117</td>\n",
       "      <td>...</td>\n",
       "      <td>88</td>\n",
       "      <td>111</td>\n",
       "      <td>101</td>\n",
       "      <td>44</td>\n",
       "      <td>17</td>\n",
       "      <td>114</td>\n",
       "      <td>86</td>\n",
       "      <td>39</td>\n",
       "      <td>245</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499326780</th>\n",
       "      <td>64</td>\n",
       "      <td>25</td>\n",
       "      <td>412</td>\n",
       "      <td>250</td>\n",
       "      <td>87</td>\n",
       "      <td>75</td>\n",
       "      <td>73</td>\n",
       "      <td>86</td>\n",
       "      <td>46</td>\n",
       "      <td>48</td>\n",
       "      <td>...</td>\n",
       "      <td>99</td>\n",
       "      <td>123</td>\n",
       "      <td>110</td>\n",
       "      <td>62</td>\n",
       "      <td>75</td>\n",
       "      <td>117</td>\n",
       "      <td>71</td>\n",
       "      <td>59</td>\n",
       "      <td>114</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499328351</th>\n",
       "      <td>164</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>116</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>189</td>\n",
       "      <td>100</td>\n",
       "      <td>98</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>166</td>\n",
       "      <td>81</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499328366</th>\n",
       "      <td>36</td>\n",
       "      <td>12</td>\n",
       "      <td>82</td>\n",
       "      <td>52</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>38</td>\n",
       "      <td>34</td>\n",
       "      <td>15</td>\n",
       "      <td>10</td>\n",
       "      <td>37</td>\n",
       "      <td>42</td>\n",
       "      <td>27</td>\n",
       "      <td>29</td>\n",
       "      <td>44</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499329195</th>\n",
       "      <td>90</td>\n",
       "      <td>155</td>\n",
       "      <td>168</td>\n",
       "      <td>114</td>\n",
       "      <td>165</td>\n",
       "      <td>115</td>\n",
       "      <td>120</td>\n",
       "      <td>54</td>\n",
       "      <td>146</td>\n",
       "      <td>186</td>\n",
       "      <td>...</td>\n",
       "      <td>50</td>\n",
       "      <td>62</td>\n",
       "      <td>262</td>\n",
       "      <td>233</td>\n",
       "      <td>333</td>\n",
       "      <td>83</td>\n",
       "      <td>61</td>\n",
       "      <td>177</td>\n",
       "      <td>115</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499330118</th>\n",
       "      <td>55</td>\n",
       "      <td>12</td>\n",
       "      <td>137</td>\n",
       "      <td>92</td>\n",
       "      <td>103</td>\n",
       "      <td>80</td>\n",
       "      <td>77</td>\n",
       "      <td>116</td>\n",
       "      <td>21</td>\n",
       "      <td>88</td>\n",
       "      <td>...</td>\n",
       "      <td>56</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>13</td>\n",
       "      <td>58</td>\n",
       "      <td>98</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499330167</th>\n",
       "      <td>97</td>\n",
       "      <td>29</td>\n",
       "      <td>40</td>\n",
       "      <td>493</td>\n",
       "      <td>95</td>\n",
       "      <td>266</td>\n",
       "      <td>279</td>\n",
       "      <td>515</td>\n",
       "      <td>228</td>\n",
       "      <td>181</td>\n",
       "      <td>...</td>\n",
       "      <td>434</td>\n",
       "      <td>783</td>\n",
       "      <td>268</td>\n",
       "      <td>267</td>\n",
       "      <td>82</td>\n",
       "      <td>627</td>\n",
       "      <td>412</td>\n",
       "      <td>341</td>\n",
       "      <td>310</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499334295</th>\n",
       "      <td>168</td>\n",
       "      <td>174</td>\n",
       "      <td>111</td>\n",
       "      <td>136</td>\n",
       "      <td>200</td>\n",
       "      <td>190</td>\n",
       "      <td>163</td>\n",
       "      <td>133</td>\n",
       "      <td>224</td>\n",
       "      <td>285</td>\n",
       "      <td>...</td>\n",
       "      <td>181</td>\n",
       "      <td>154</td>\n",
       "      <td>228</td>\n",
       "      <td>316</td>\n",
       "      <td>348</td>\n",
       "      <td>178</td>\n",
       "      <td>174</td>\n",
       "      <td>258</td>\n",
       "      <td>166</td>\n",
       "      <td>208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499334626</th>\n",
       "      <td>92</td>\n",
       "      <td>292</td>\n",
       "      <td>330</td>\n",
       "      <td>130</td>\n",
       "      <td>115</td>\n",
       "      <td>152</td>\n",
       "      <td>144</td>\n",
       "      <td>101</td>\n",
       "      <td>219</td>\n",
       "      <td>230</td>\n",
       "      <td>...</td>\n",
       "      <td>165</td>\n",
       "      <td>129</td>\n",
       "      <td>335</td>\n",
       "      <td>439</td>\n",
       "      <td>256</td>\n",
       "      <td>189</td>\n",
       "      <td>193</td>\n",
       "      <td>240</td>\n",
       "      <td>163</td>\n",
       "      <td>209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499334792</th>\n",
       "      <td>805</td>\n",
       "      <td>671</td>\n",
       "      <td>539</td>\n",
       "      <td>612</td>\n",
       "      <td>887</td>\n",
       "      <td>942</td>\n",
       "      <td>792</td>\n",
       "      <td>559</td>\n",
       "      <td>1002</td>\n",
       "      <td>1018</td>\n",
       "      <td>...</td>\n",
       "      <td>752</td>\n",
       "      <td>521</td>\n",
       "      <td>824</td>\n",
       "      <td>1328</td>\n",
       "      <td>1467</td>\n",
       "      <td>745</td>\n",
       "      <td>777</td>\n",
       "      <td>1104</td>\n",
       "      <td>801</td>\n",
       "      <td>863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499335939</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499336206</th>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>27</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "      <td>19</td>\n",
       "      <td>23</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>40</td>\n",
       "      <td>34</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>34</td>\n",
       "      <td>26</td>\n",
       "      <td>8</td>\n",
       "      <td>17</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499336276</th>\n",
       "      <td>89</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>171</td>\n",
       "      <td>141</td>\n",
       "      <td>146</td>\n",
       "      <td>122</td>\n",
       "      <td>174</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499336992</th>\n",
       "      <td>548</td>\n",
       "      <td>374</td>\n",
       "      <td>274</td>\n",
       "      <td>457</td>\n",
       "      <td>651</td>\n",
       "      <td>541</td>\n",
       "      <td>497</td>\n",
       "      <td>405</td>\n",
       "      <td>633</td>\n",
       "      <td>789</td>\n",
       "      <td>...</td>\n",
       "      <td>335</td>\n",
       "      <td>288</td>\n",
       "      <td>508</td>\n",
       "      <td>625</td>\n",
       "      <td>1211</td>\n",
       "      <td>302</td>\n",
       "      <td>333</td>\n",
       "      <td>538</td>\n",
       "      <td>308</td>\n",
       "      <td>325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499342049</th>\n",
       "      <td>98</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>105</td>\n",
       "      <td>136</td>\n",
       "      <td>130</td>\n",
       "      <td>112</td>\n",
       "      <td>73</td>\n",
       "      <td>177</td>\n",
       "      <td>234</td>\n",
       "      <td>...</td>\n",
       "      <td>85</td>\n",
       "      <td>72</td>\n",
       "      <td>90</td>\n",
       "      <td>165</td>\n",
       "      <td>148</td>\n",
       "      <td>83</td>\n",
       "      <td>97</td>\n",
       "      <td>198</td>\n",
       "      <td>128</td>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499343767</th>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499343769</th>\n",
       "      <td>30</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>21</td>\n",
       "      <td>17</td>\n",
       "      <td>11</td>\n",
       "      <td>14</td>\n",
       "      <td>25</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>19</td>\n",
       "      <td>17</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499347240</th>\n",
       "      <td>23</td>\n",
       "      <td>24</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>13</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>26</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>16</td>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499348546</th>\n",
       "      <td>497</td>\n",
       "      <td>664</td>\n",
       "      <td>779</td>\n",
       "      <td>575</td>\n",
       "      <td>891</td>\n",
       "      <td>873</td>\n",
       "      <td>698</td>\n",
       "      <td>483</td>\n",
       "      <td>746</td>\n",
       "      <td>965</td>\n",
       "      <td>...</td>\n",
       "      <td>470</td>\n",
       "      <td>438</td>\n",
       "      <td>681</td>\n",
       "      <td>1164</td>\n",
       "      <td>825</td>\n",
       "      <td>543</td>\n",
       "      <td>532</td>\n",
       "      <td>1101</td>\n",
       "      <td>575</td>\n",
       "      <td>659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499348654</th>\n",
       "      <td>29</td>\n",
       "      <td>34</td>\n",
       "      <td>32</td>\n",
       "      <td>16</td>\n",
       "      <td>30</td>\n",
       "      <td>44</td>\n",
       "      <td>26</td>\n",
       "      <td>35</td>\n",
       "      <td>26</td>\n",
       "      <td>50</td>\n",
       "      <td>...</td>\n",
       "      <td>34</td>\n",
       "      <td>36</td>\n",
       "      <td>62</td>\n",
       "      <td>53</td>\n",
       "      <td>45</td>\n",
       "      <td>39</td>\n",
       "      <td>24</td>\n",
       "      <td>55</td>\n",
       "      <td>26</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499350441</th>\n",
       "      <td>60</td>\n",
       "      <td>100</td>\n",
       "      <td>54</td>\n",
       "      <td>31</td>\n",
       "      <td>37</td>\n",
       "      <td>49</td>\n",
       "      <td>30</td>\n",
       "      <td>29</td>\n",
       "      <td>65</td>\n",
       "      <td>77</td>\n",
       "      <td>...</td>\n",
       "      <td>34</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>88</td>\n",
       "      <td>69</td>\n",
       "      <td>32</td>\n",
       "      <td>42</td>\n",
       "      <td>54</td>\n",
       "      <td>53</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499352783</th>\n",
       "      <td>30</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>70</td>\n",
       "      <td>16</td>\n",
       "      <td>55</td>\n",
       "      <td>66</td>\n",
       "      <td>76</td>\n",
       "      <td>38</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>107</td>\n",
       "      <td>90</td>\n",
       "      <td>41</td>\n",
       "      <td>29</td>\n",
       "      <td>19</td>\n",
       "      <td>87</td>\n",
       "      <td>82</td>\n",
       "      <td>28</td>\n",
       "      <td>23</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32 rows × 377 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "rnaseq_profile_id  488395315  496100277  496100278  496100279  496100281   \n",
       "gene_id_mapped                                                             \n",
       "499309627                 21         85         83         44         40  \\\n",
       "499315162                 51        143        275         70        112   \n",
       "499315537                 13         22         14         49         28   \n",
       "499315843                 18         13          6         12          5   \n",
       "499317231                 31         15         30         42         55   \n",
       "499317260                  4          3          6         10          4   \n",
       "499319747                  7          5          2         13          3   \n",
       "499322427                638        216        184        806        435   \n",
       "499323096                102        235        205        136        244   \n",
       "499323166                462        362        297        466        545   \n",
       "499324505                 11          7          3        108         43   \n",
       "499326780                 64         25        412        250         87   \n",
       "499328351                164         43          0          0          0   \n",
       "499328366                 36         12         82         52          8   \n",
       "499329195                 90        155        168        114        165   \n",
       "499330118                 55         12        137         92        103   \n",
       "499330167                 97         29         40        493         95   \n",
       "499334295                168        174        111        136        200   \n",
       "499334626                 92        292        330        130        115   \n",
       "499334792                805        671        539        612        887   \n",
       "499335939                  4          0          1          1          1   \n",
       "499336206                 11          3          1         27         12   \n",
       "499336276                 89          0         45        171        141   \n",
       "499336992                548        374        274        457        651   \n",
       "499342049                 98        100         90        105        136   \n",
       "499343767                 14          2          1          8          9   \n",
       "499343769                 30          3          5          6         11   \n",
       "499347240                 23         24          7          5         13   \n",
       "499348546                497        664        779        575        891   \n",
       "499348654                 29         34         32         16         30   \n",
       "499350441                 60        100         54         31         37   \n",
       "499352783                 30         12         13         70         16   \n",
       "\n",
       "rnaseq_profile_id  496100283  496100284  496100285  496100287  496100288  ...   \n",
       "gene_id_mapped                                                            ...   \n",
       "499309627                 34         39         32         26         27  ...  \\\n",
       "499315162                 80         57         45         28        100  ...   \n",
       "499315537                 27         48         44         14         48  ...   \n",
       "499315843                 25         18         15         20         19  ...   \n",
       "499317231                 38         35         62         37         39  ...   \n",
       "499317260                  4         10         10          6          4  ...   \n",
       "499319747                  9         11         16          5          6  ...   \n",
       "499322427                222        287        713        491        513  ...   \n",
       "499323096                204        191        132        157        192  ...   \n",
       "499323166                725        580        373        586        904  ...   \n",
       "499324505                 68         59        119         45        117  ...   \n",
       "499326780                 75         73         86         46         48  ...   \n",
       "499328351                116         37          0          0          0  ...   \n",
       "499328366                  9          9          7         10          7  ...   \n",
       "499329195                115        120         54        146        186  ...   \n",
       "499330118                 80         77        116         21         88  ...   \n",
       "499330167                266        279        515        228        181  ...   \n",
       "499334295                190        163        133        224        285  ...   \n",
       "499334626                152        144        101        219        230  ...   \n",
       "499334792                942        792        559       1002       1018  ...   \n",
       "499335939                  2          3          1          5          3  ...   \n",
       "499336206                 15         19         23         21          4  ...   \n",
       "499336276                146        122        174          0          0  ...   \n",
       "499336992                541        497        405        633        789  ...   \n",
       "499342049                130        112         73        177        234  ...   \n",
       "499343767                 10          7          5          7          1  ...   \n",
       "499343769                 21         17         11         14         25  ...   \n",
       "499347240                 17         17          3         13         26  ...   \n",
       "499348546                873        698        483        746        965  ...   \n",
       "499348654                 44         26         35         26         50  ...   \n",
       "499350441                 49         30         29         65         77  ...   \n",
       "499352783                 55         66         76         38          8  ...   \n",
       "\n",
       "rnaseq_profile_id  496100661  496100663  496100664  496100665  496100666   \n",
       "gene_id_mapped                                                             \n",
       "499309627                 43         45         24         20         17  \\\n",
       "499315162                 76         95        151         28        105   \n",
       "499315537                 70         43         51          9         24   \n",
       "499315843                 26         12         11         39         20   \n",
       "499317231                 64         89         47         23         22   \n",
       "499317260                 14         13          5          5          6   \n",
       "499319747                 39         24          8          5          3   \n",
       "499322427                702       1357       1418        308        976   \n",
       "499323096                117        106        197        261        225   \n",
       "499323166                464        372        519        938        606   \n",
       "499324505                 88        111        101         44         17   \n",
       "499326780                 99        123        110         62         75   \n",
       "499328351                  0        189        100         98          4   \n",
       "499328366                 38         34         15         10         37   \n",
       "499329195                 50         62        262        233        333   \n",
       "499330118                 56         11          0         11          0   \n",
       "499330167                434        783        268        267         82   \n",
       "499334295                181        154        228        316        348   \n",
       "499334626                165        129        335        439        256   \n",
       "499334792                752        521        824       1328       1467   \n",
       "499335939                  6          0          3          0          0   \n",
       "499336206                 40         34         14          6          7   \n",
       "499336276                  0          0          0          0          0   \n",
       "499336992                335        288        508        625       1211   \n",
       "499342049                 85         72         90        165        148   \n",
       "499343767                  5          5          4         16         13   \n",
       "499343769                  8          7         10         19         17   \n",
       "499347240                  4          1         12         23         23   \n",
       "499348546                470        438        681       1164        825   \n",
       "499348654                 34         36         62         53         45   \n",
       "499350441                 34         30         72         88         69   \n",
       "499352783                107         90         41         29         19   \n",
       "\n",
       "rnaseq_profile_id  496100667  496100669  496100670  496100671  496100672  \n",
       "gene_id_mapped                                                            \n",
       "499309627                 53         52         26         56         40  \n",
       "499315162                 87         78         66        117         39  \n",
       "499315537                103        103         11        124         41  \n",
       "499315843                 23         29         35         11         10  \n",
       "499317231                 73         42         48         46         41  \n",
       "499317260                  8          4          0         11          7  \n",
       "499319747                 39         26          6         21         10  \n",
       "499322427               1059        681        512        648        742  \n",
       "499323096                129        146        250        169        183  \n",
       "499323166                526        508        859        620        547  \n",
       "499324505                114         86         39        245         27  \n",
       "499326780                117         71         59        114        185  \n",
       "499328351                  0          0        166         81          0  \n",
       "499328366                 42         27         29         44          5  \n",
       "499329195                 83         61        177        115        129  \n",
       "499330118                 72         13         58         98         91  \n",
       "499330167                627        412        341        310        103  \n",
       "499334295                178        174        258        166        208  \n",
       "499334626                189        193        240        163        209  \n",
       "499334792                745        777       1104        801        863  \n",
       "499335939                  6         11          2          4          5  \n",
       "499336206                 34         26          8         17         16  \n",
       "499336276                  0          0          5          0          0  \n",
       "499336992                302        333        538        308        325  \n",
       "499342049                 83         97        198        128        113  \n",
       "499343767                  5          4         19          2          2  \n",
       "499343769                  7          6          9          6         18  \n",
       "499347240                  6          9         16         10         12  \n",
       "499348546                543        532       1101        575        659  \n",
       "499348654                 39         24         55         26         57  \n",
       "499350441                 32         42         54         53         49  \n",
       "499352783                 87         82         28         23         50  \n",
       "\n",
       "[32 rows x 377 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter by top genes (top and bottom 10)\n",
    "ml_df = counts[counts.index.isin(gene_list)].copy()\n",
    "ml_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rnaseq_profile_id</th>\n",
       "      <th>499309627</th>\n",
       "      <th>499315162</th>\n",
       "      <th>499315537</th>\n",
       "      <th>499315843</th>\n",
       "      <th>499317231</th>\n",
       "      <th>499317260</th>\n",
       "      <th>499319747</th>\n",
       "      <th>499322427</th>\n",
       "      <th>499323096</th>\n",
       "      <th>...</th>\n",
       "      <th>499336992</th>\n",
       "      <th>499342049</th>\n",
       "      <th>499343767</th>\n",
       "      <th>499343769</th>\n",
       "      <th>499347240</th>\n",
       "      <th>499348546</th>\n",
       "      <th>499348654</th>\n",
       "      <th>499350441</th>\n",
       "      <th>499352783</th>\n",
       "      <th>Condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>488395315</td>\n",
       "      <td>21</td>\n",
       "      <td>51</td>\n",
       "      <td>13</td>\n",
       "      <td>18</td>\n",
       "      <td>31</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>638</td>\n",
       "      <td>102</td>\n",
       "      <td>...</td>\n",
       "      <td>548</td>\n",
       "      <td>98</td>\n",
       "      <td>14</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "      <td>497</td>\n",
       "      <td>29</td>\n",
       "      <td>60</td>\n",
       "      <td>30</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>496100277</td>\n",
       "      <td>85</td>\n",
       "      <td>143</td>\n",
       "      <td>22</td>\n",
       "      <td>13</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>216</td>\n",
       "      <td>235</td>\n",
       "      <td>...</td>\n",
       "      <td>374</td>\n",
       "      <td>100</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>24</td>\n",
       "      <td>664</td>\n",
       "      <td>34</td>\n",
       "      <td>100</td>\n",
       "      <td>12</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>496100278</td>\n",
       "      <td>83</td>\n",
       "      <td>275</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>30</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>184</td>\n",
       "      <td>205</td>\n",
       "      <td>...</td>\n",
       "      <td>274</td>\n",
       "      <td>90</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>779</td>\n",
       "      <td>32</td>\n",
       "      <td>54</td>\n",
       "      <td>13</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>496100279</td>\n",
       "      <td>44</td>\n",
       "      <td>70</td>\n",
       "      <td>49</td>\n",
       "      <td>12</td>\n",
       "      <td>42</td>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "      <td>806</td>\n",
       "      <td>136</td>\n",
       "      <td>...</td>\n",
       "      <td>457</td>\n",
       "      <td>105</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>575</td>\n",
       "      <td>16</td>\n",
       "      <td>31</td>\n",
       "      <td>70</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>496100281</td>\n",
       "      <td>40</td>\n",
       "      <td>112</td>\n",
       "      <td>28</td>\n",
       "      <td>5</td>\n",
       "      <td>55</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>435</td>\n",
       "      <td>244</td>\n",
       "      <td>...</td>\n",
       "      <td>651</td>\n",
       "      <td>136</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>891</td>\n",
       "      <td>30</td>\n",
       "      <td>37</td>\n",
       "      <td>16</td>\n",
       "      <td>dementia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>496100667</td>\n",
       "      <td>53</td>\n",
       "      <td>87</td>\n",
       "      <td>103</td>\n",
       "      <td>23</td>\n",
       "      <td>73</td>\n",
       "      <td>8</td>\n",
       "      <td>39</td>\n",
       "      <td>1059</td>\n",
       "      <td>129</td>\n",
       "      <td>...</td>\n",
       "      <td>302</td>\n",
       "      <td>83</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>543</td>\n",
       "      <td>39</td>\n",
       "      <td>32</td>\n",
       "      <td>87</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>496100669</td>\n",
       "      <td>52</td>\n",
       "      <td>78</td>\n",
       "      <td>103</td>\n",
       "      <td>29</td>\n",
       "      <td>42</td>\n",
       "      <td>4</td>\n",
       "      <td>26</td>\n",
       "      <td>681</td>\n",
       "      <td>146</td>\n",
       "      <td>...</td>\n",
       "      <td>333</td>\n",
       "      <td>97</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>532</td>\n",
       "      <td>24</td>\n",
       "      <td>42</td>\n",
       "      <td>82</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>496100670</td>\n",
       "      <td>26</td>\n",
       "      <td>66</td>\n",
       "      <td>11</td>\n",
       "      <td>35</td>\n",
       "      <td>48</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>512</td>\n",
       "      <td>250</td>\n",
       "      <td>...</td>\n",
       "      <td>538</td>\n",
       "      <td>198</td>\n",
       "      <td>19</td>\n",
       "      <td>9</td>\n",
       "      <td>16</td>\n",
       "      <td>1101</td>\n",
       "      <td>55</td>\n",
       "      <td>54</td>\n",
       "      <td>28</td>\n",
       "      <td>dementia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>496100671</td>\n",
       "      <td>56</td>\n",
       "      <td>117</td>\n",
       "      <td>124</td>\n",
       "      <td>11</td>\n",
       "      <td>46</td>\n",
       "      <td>11</td>\n",
       "      <td>21</td>\n",
       "      <td>648</td>\n",
       "      <td>169</td>\n",
       "      <td>...</td>\n",
       "      <td>308</td>\n",
       "      <td>128</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>575</td>\n",
       "      <td>26</td>\n",
       "      <td>53</td>\n",
       "      <td>23</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>496100672</td>\n",
       "      <td>40</td>\n",
       "      <td>39</td>\n",
       "      <td>41</td>\n",
       "      <td>10</td>\n",
       "      <td>41</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>742</td>\n",
       "      <td>183</td>\n",
       "      <td>...</td>\n",
       "      <td>325</td>\n",
       "      <td>113</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>659</td>\n",
       "      <td>57</td>\n",
       "      <td>49</td>\n",
       "      <td>50</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>377 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     rnaseq_profile_id  499309627  499315162  499315537  499315843  499317231   \n",
       "0            488395315         21         51         13         18         31  \\\n",
       "1            496100277         85        143         22         13         15   \n",
       "2            496100278         83        275         14          6         30   \n",
       "3            496100279         44         70         49         12         42   \n",
       "4            496100281         40        112         28          5         55   \n",
       "..                 ...        ...        ...        ...        ...        ...   \n",
       "372          496100667         53         87        103         23         73   \n",
       "373          496100669         52         78        103         29         42   \n",
       "374          496100670         26         66         11         35         48   \n",
       "375          496100671         56        117        124         11         46   \n",
       "376          496100672         40         39         41         10         41   \n",
       "\n",
       "     499317260  499319747  499322427  499323096  ...  499336992  499342049   \n",
       "0            4          7        638        102  ...        548         98  \\\n",
       "1            3          5        216        235  ...        374        100   \n",
       "2            6          2        184        205  ...        274         90   \n",
       "3           10         13        806        136  ...        457        105   \n",
       "4            4          3        435        244  ...        651        136   \n",
       "..         ...        ...        ...        ...  ...        ...        ...   \n",
       "372          8         39       1059        129  ...        302         83   \n",
       "373          4         26        681        146  ...        333         97   \n",
       "374          0          6        512        250  ...        538        198   \n",
       "375         11         21        648        169  ...        308        128   \n",
       "376          7         10        742        183  ...        325        113   \n",
       "\n",
       "     499343767  499343769  499347240  499348546  499348654  499350441   \n",
       "0           14         30         23        497         29         60  \\\n",
       "1            2          3         24        664         34        100   \n",
       "2            1          5          7        779         32         54   \n",
       "3            8          6          5        575         16         31   \n",
       "4            9         11         13        891         30         37   \n",
       "..         ...        ...        ...        ...        ...        ...   \n",
       "372          5          7          6        543         39         32   \n",
       "373          4          6          9        532         24         42   \n",
       "374         19          9         16       1101         55         54   \n",
       "375          2          6         10        575         26         53   \n",
       "376          2         18         12        659         57         49   \n",
       "\n",
       "     499352783  Condition  \n",
       "0           30    control  \n",
       "1           12    control  \n",
       "2           13    control  \n",
       "3           70    control  \n",
       "4           16   dementia  \n",
       "..         ...        ...  \n",
       "372         87    control  \n",
       "373         82    control  \n",
       "374         28   dementia  \n",
       "375         23    control  \n",
       "376         50    control  \n",
       "\n",
       "[377 rows x 34 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Transform and apply conditions (y_label)\n",
    "ml_df = ml_df.T.reset_index().rename_axis(None, axis = 1)\n",
    "ml_df['Condition'] = samples['Condition']\n",
    "ml_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(190, 80, 107)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#samples (rna_profile_ids) by donor data splt\n",
    "train_samples = samples[samples['donor_id'].isin(train_ids)]['rnaseq_profile_id']\n",
    "val_samples = samples[samples['donor_id'].isin(validate_ids)]['rnaseq_profile_id']\n",
    "test_samples = samples[samples['donor_id'].isin(test_ids)]['rnaseq_profile_id']\n",
    "\n",
    "len(train_samples), len(val_samples), len(test_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now can filter by train, val, test, split\n",
    "train_df = ml_df[ml_df['rnaseq_profile_id'].isin(train_samples)].drop(columns = 'rnaseq_profile_id')\n",
    "val_df = ml_df[ml_df['rnaseq_profile_id'].isin(val_samples)].drop(columns = 'rnaseq_profile_id')\n",
    "test_df = ml_df[ml_df['rnaseq_profile_id'].isin(test_samples)].drop(columns = 'rnaseq_profile_id')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>499309627</th>\n",
       "      <th>499315162</th>\n",
       "      <th>499315537</th>\n",
       "      <th>499315843</th>\n",
       "      <th>499317231</th>\n",
       "      <th>499317260</th>\n",
       "      <th>499319747</th>\n",
       "      <th>499322427</th>\n",
       "      <th>499323096</th>\n",
       "      <th>499323166</th>\n",
       "      <th>...</th>\n",
       "      <th>499336992</th>\n",
       "      <th>499342049</th>\n",
       "      <th>499343767</th>\n",
       "      <th>499343769</th>\n",
       "      <th>499347240</th>\n",
       "      <th>499348546</th>\n",
       "      <th>499348654</th>\n",
       "      <th>499350441</th>\n",
       "      <th>499352783</th>\n",
       "      <th>Condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>85</td>\n",
       "      <td>143</td>\n",
       "      <td>22</td>\n",
       "      <td>13</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>216</td>\n",
       "      <td>235</td>\n",
       "      <td>362</td>\n",
       "      <td>...</td>\n",
       "      <td>374</td>\n",
       "      <td>100</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>24</td>\n",
       "      <td>664</td>\n",
       "      <td>34</td>\n",
       "      <td>100</td>\n",
       "      <td>12</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>83</td>\n",
       "      <td>275</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>30</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>184</td>\n",
       "      <td>205</td>\n",
       "      <td>297</td>\n",
       "      <td>...</td>\n",
       "      <td>274</td>\n",
       "      <td>90</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>779</td>\n",
       "      <td>32</td>\n",
       "      <td>54</td>\n",
       "      <td>13</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>44</td>\n",
       "      <td>70</td>\n",
       "      <td>49</td>\n",
       "      <td>12</td>\n",
       "      <td>42</td>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "      <td>806</td>\n",
       "      <td>136</td>\n",
       "      <td>466</td>\n",
       "      <td>...</td>\n",
       "      <td>457</td>\n",
       "      <td>105</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>575</td>\n",
       "      <td>16</td>\n",
       "      <td>31</td>\n",
       "      <td>70</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>40</td>\n",
       "      <td>112</td>\n",
       "      <td>28</td>\n",
       "      <td>5</td>\n",
       "      <td>55</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>435</td>\n",
       "      <td>244</td>\n",
       "      <td>545</td>\n",
       "      <td>...</td>\n",
       "      <td>651</td>\n",
       "      <td>136</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>891</td>\n",
       "      <td>30</td>\n",
       "      <td>37</td>\n",
       "      <td>16</td>\n",
       "      <td>dementia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>34</td>\n",
       "      <td>80</td>\n",
       "      <td>27</td>\n",
       "      <td>25</td>\n",
       "      <td>38</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>222</td>\n",
       "      <td>204</td>\n",
       "      <td>725</td>\n",
       "      <td>...</td>\n",
       "      <td>541</td>\n",
       "      <td>130</td>\n",
       "      <td>10</td>\n",
       "      <td>21</td>\n",
       "      <td>17</td>\n",
       "      <td>873</td>\n",
       "      <td>44</td>\n",
       "      <td>49</td>\n",
       "      <td>55</td>\n",
       "      <td>dementia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>24</td>\n",
       "      <td>151</td>\n",
       "      <td>51</td>\n",
       "      <td>11</td>\n",
       "      <td>47</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>1418</td>\n",
       "      <td>197</td>\n",
       "      <td>519</td>\n",
       "      <td>...</td>\n",
       "      <td>508</td>\n",
       "      <td>90</td>\n",
       "      <td>4</td>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>681</td>\n",
       "      <td>62</td>\n",
       "      <td>72</td>\n",
       "      <td>41</td>\n",
       "      <td>dementia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>53</td>\n",
       "      <td>87</td>\n",
       "      <td>103</td>\n",
       "      <td>23</td>\n",
       "      <td>73</td>\n",
       "      <td>8</td>\n",
       "      <td>39</td>\n",
       "      <td>1059</td>\n",
       "      <td>129</td>\n",
       "      <td>526</td>\n",
       "      <td>...</td>\n",
       "      <td>302</td>\n",
       "      <td>83</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>543</td>\n",
       "      <td>39</td>\n",
       "      <td>32</td>\n",
       "      <td>87</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>52</td>\n",
       "      <td>78</td>\n",
       "      <td>103</td>\n",
       "      <td>29</td>\n",
       "      <td>42</td>\n",
       "      <td>4</td>\n",
       "      <td>26</td>\n",
       "      <td>681</td>\n",
       "      <td>146</td>\n",
       "      <td>508</td>\n",
       "      <td>...</td>\n",
       "      <td>333</td>\n",
       "      <td>97</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>532</td>\n",
       "      <td>24</td>\n",
       "      <td>42</td>\n",
       "      <td>82</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>56</td>\n",
       "      <td>117</td>\n",
       "      <td>124</td>\n",
       "      <td>11</td>\n",
       "      <td>46</td>\n",
       "      <td>11</td>\n",
       "      <td>21</td>\n",
       "      <td>648</td>\n",
       "      <td>169</td>\n",
       "      <td>620</td>\n",
       "      <td>...</td>\n",
       "      <td>308</td>\n",
       "      <td>128</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>575</td>\n",
       "      <td>26</td>\n",
       "      <td>53</td>\n",
       "      <td>23</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>40</td>\n",
       "      <td>39</td>\n",
       "      <td>41</td>\n",
       "      <td>10</td>\n",
       "      <td>41</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>742</td>\n",
       "      <td>183</td>\n",
       "      <td>547</td>\n",
       "      <td>...</td>\n",
       "      <td>325</td>\n",
       "      <td>113</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>659</td>\n",
       "      <td>57</td>\n",
       "      <td>49</td>\n",
       "      <td>50</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     499309627  499315162  499315537  499315843  499317231  499317260   \n",
       "1           85        143         22         13         15          3  \\\n",
       "2           83        275         14          6         30          6   \n",
       "3           44         70         49         12         42         10   \n",
       "4           40        112         28          5         55          4   \n",
       "5           34         80         27         25         38          4   \n",
       "..         ...        ...        ...        ...        ...        ...   \n",
       "369         24        151         51         11         47          5   \n",
       "372         53         87        103         23         73          8   \n",
       "373         52         78        103         29         42          4   \n",
       "375         56        117        124         11         46         11   \n",
       "376         40         39         41         10         41          7   \n",
       "\n",
       "     499319747  499322427  499323096  499323166  ...  499336992  499342049   \n",
       "1            5        216        235        362  ...        374        100  \\\n",
       "2            2        184        205        297  ...        274         90   \n",
       "3           13        806        136        466  ...        457        105   \n",
       "4            3        435        244        545  ...        651        136   \n",
       "5            9        222        204        725  ...        541        130   \n",
       "..         ...        ...        ...        ...  ...        ...        ...   \n",
       "369          8       1418        197        519  ...        508         90   \n",
       "372         39       1059        129        526  ...        302         83   \n",
       "373         26        681        146        508  ...        333         97   \n",
       "375         21        648        169        620  ...        308        128   \n",
       "376         10        742        183        547  ...        325        113   \n",
       "\n",
       "     499343767  499343769  499347240  499348546  499348654  499350441   \n",
       "1            2          3         24        664         34        100  \\\n",
       "2            1          5          7        779         32         54   \n",
       "3            8          6          5        575         16         31   \n",
       "4            9         11         13        891         30         37   \n",
       "5           10         21         17        873         44         49   \n",
       "..         ...        ...        ...        ...        ...        ...   \n",
       "369          4         10         12        681         62         72   \n",
       "372          5          7          6        543         39         32   \n",
       "373          4          6          9        532         24         42   \n",
       "375          2          6         10        575         26         53   \n",
       "376          2         18         12        659         57         49   \n",
       "\n",
       "     499352783  Condition  \n",
       "1           12    control  \n",
       "2           13    control  \n",
       "3           70    control  \n",
       "4           16   dementia  \n",
       "5           55   dementia  \n",
       "..         ...        ...  \n",
       "369         41   dementia  \n",
       "372         87    control  \n",
       "373         82    control  \n",
       "375         23    control  \n",
       "376         50    control  \n",
       "\n",
       "[190 rows x 33 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# quick check\n",
    "train_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Base Models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train = train_df.drop(columns='Condition')\n",
    "y_train = train_df['Condition']\n",
    "\n",
    "X_val = val_df.drop(columns='Condition')\n",
    "y_val = val_df['Condition']\n",
    "\n",
    "# Scale data and transform data\n",
    "scaler = StandardScaler()\n",
    "scaler.fit(X_train)\n",
    "X_train_= scaler.transform(X_train)\n",
    "X_val_= scaler.transform(X_val)\n",
    "\n",
    "y_train = y_train.apply(lambda x: 1 if x == 'dementia' else 0)\n",
    "y_val = y_val.apply(lambda x: 1 if  x=='dementia' else 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.40      1.00      0.57        32\n",
      "           1       0.00      0.00      0.00        48\n",
      "\n",
      "    accuracy                           0.40        80\n",
      "   macro avg       0.20      0.50      0.29        80\n",
      "weighted avg       0.16      0.40      0.23        80\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/homebrew/Caskroom/miniforge/base/envs/mads/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n",
      "/opt/homebrew/Caskroom/miniforge/base/envs/mads/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n",
      "/opt/homebrew/Caskroom/miniforge/base/envs/mads/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n"
     ]
    }
   ],
   "source": [
    "dummy = DummyClassifier(strategy= 'most_frequent', random_state=42)\n",
    "dummy.fit(X_train_, y_train)\n",
    "preds = dummy.predict(X_val_)\n",
    "score = classification_report(y_val, preds)\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.59      0.72      0.65        32\n",
      "           1       0.78      0.67      0.72        48\n",
      "\n",
      "    accuracy                           0.69        80\n",
      "   macro avg       0.69      0.69      0.68        80\n",
      "weighted avg       0.70      0.69      0.69        80\n",
      "\n"
     ]
    }
   ],
   "source": [
    "reg = LogisticRegression(random_state=42)\n",
    "reg.fit(X_train_, y_train)\n",
    "preds = reg.predict(X_val_)\n",
    "score = classification_report(y_val, preds)\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.53      0.81      0.64        32\n",
      "           1       0.81      0.52      0.63        48\n",
      "\n",
      "    accuracy                           0.64        80\n",
      "   macro avg       0.67      0.67      0.64        80\n",
      "weighted avg       0.70      0.64      0.64        80\n",
      "\n"
     ]
    }
   ],
   "source": [
    "tree = DecisionTreeClassifier(random_state = 42)\n",
    "tree.fit(X_train_, y_train)\n",
    "preds = tree.predict(X_val_)\n",
    "score = classification_report(y_val, preds)\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.56      0.78      0.65        32\n",
      "           1       0.80      0.58      0.67        48\n",
      "\n",
      "    accuracy                           0.66        80\n",
      "   macro avg       0.68      0.68      0.66        80\n",
      "weighted avg       0.70      0.66      0.66        80\n",
      "\n"
     ]
    }
   ],
   "source": [
    "forest = RandomForestClassifier(random_state= 42)\n",
    "forest.fit(X_train_, y_train)\n",
    "preds = forest.predict(X_val_)\n",
    "score = classification_report(y_val, preds)\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.58      0.69      0.63        32\n",
      "           1       0.76      0.67      0.71        48\n",
      "\n",
      "    accuracy                           0.68        80\n",
      "   macro avg       0.67      0.68      0.67        80\n",
      "weighted avg       0.69      0.68      0.68        80\n",
      "\n"
     ]
    }
   ],
   "source": [
    "linear_svm = SVC(kernel='linear', random_state=42)\n",
    "linear_svm.fit(X_train_, y_train)\n",
    "preds = linear_svm.predict(X_val_)\n",
    "score = classification_report(y_val, preds)\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.62      0.78      0.69        32\n",
      "           1       0.82      0.69      0.75        48\n",
      "\n",
      "    accuracy                           0.73        80\n",
      "   macro avg       0.72      0.73      0.72        80\n",
      "weighted avg       0.74      0.72      0.73        80\n",
      "\n"
     ]
    }
   ],
   "source": [
    "rbf_svm = SVC(kernel='rbf', random_state=42)\n",
    "rbf_svm.fit(X_train_, y_train)\n",
    "preds = rbf_svm.predict(X_val_)\n",
    "score = classification_report(y_val, preds)\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.61      0.78      0.68        32\n",
      "           1       0.82      0.67      0.74        48\n",
      "\n",
      "    accuracy                           0.71        80\n",
      "   macro avg       0.72      0.72      0.71        80\n",
      "weighted avg       0.74      0.71      0.72        80\n",
      "\n"
     ]
    }
   ],
   "source": [
    "boost = GradientBoostingClassifier(random_state=42)\n",
    "boost.fit(X_train_, y_train)\n",
    "preds = boost.predict(X_val_)\n",
    "score = classification_report(y_val, preds)\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.74      0.81      0.78        32\n",
      "           1       0.87      0.81      0.84        48\n",
      "\n",
      "    accuracy                           0.81        80\n",
      "   macro avg       0.80      0.81      0.81        80\n",
      "weighted avg       0.82      0.81      0.81        80\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "gauss = GaussianNB()\n",
    "gauss.fit(X_train_, y_train)\n",
    "preds = gauss.predict(X_val_)\n",
    "score = classification_report(y_val, preds)\n",
    "print(score)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.8583377221947985\n"
     ]
    }
   ],
   "source": [
    "rbf_svm = SVC(kernel='rbf', random_state=42, probability= True)\n",
    "rbf_svm.fit(X_train_, y_train)\n",
    "preds = rbf_svm.predict_proba(X_val_)\n",
    "score = average_precision_score(y_val, preds[:,1])\n",
    "print(score)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.769193004836744\n"
     ]
    }
   ],
   "source": [
    "forest = RandomForestClassifier(random_state= 42)\n",
    "forest.fit(X_train_, y_train)\n",
    "preds = forest.predict_proba(X_val_)\n",
    "score = average_precision_score(y_val, preds[:,1])\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use AVERAGE Precision score to evaluate several models\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GaussianNB          0.941\n",
       "SVM_radial          0.858\n",
       "Log_Reg             0.820\n",
       "SVM_linear          0.815\n",
       "Gradient_boosted    0.803\n",
       "RF                  0.769\n",
       "DT                  0.708\n",
       "Dummy_most_freq     0.600\n",
       "dtype: float64"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Use AVERAGE Precision score\n",
    "models = [\n",
    "    LogisticRegression(random_state=42),\n",
    "    DecisionTreeClassifier(random_state=42),\n",
    "    RandomForestClassifier(random_state=42),\n",
    "    SVC(kernel=\"linear\", random_state=42, probability=True),\n",
    "    SVC(kernel=\"rbf\", random_state=42, probability=True),\n",
    "    GaussianNB(),\n",
    "    GradientBoostingClassifier(random_state=42),\n",
    "    DummyClassifier(strategy= 'most_frequent', random_state=42),\n",
    "]\n",
    "model_names = [\"Log_Reg\", \"DT\", \"RF\", 'SVM_linear', \"SVM_radial\", \"GaussianNB\", 'Gradient_boosted', 'Dummy_most_freq']\n",
    "\n",
    "scores = {}\n",
    "saved_models = {}\n",
    "for name, model in zip(model_names, models):\n",
    "    model.fit(X_train_, y_train)\n",
    "    saved_models[name] = model\n",
    "    preds = model.predict_proba(X_val_)\n",
    "    score = average_precision_score(y_val, preds[:, 1])\n",
    "    scores[name] = score\n",
    "\n",
    "pd.Series(scores).sort_values(ascending=False).round(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_data = pd.read_csv('../data/raw/gene_expression_matrix_2016-03-03/rows-genes.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quick look at log reg coefficients\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Quick look at log reg coefficients\n",
    "support_ids = X_train.columns.to_list()\n",
    "log_reg_coefs = saved_models['Log_Reg'].coef_.tolist()[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>coef</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_entrez_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>499348546</td>\n",
       "      <td>0.637636</td>\n",
       "      <td>19</td>\n",
       "      <td>23152</td>\n",
       "      <td>CIC</td>\n",
       "      <td>capicua transcriptional repressor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>499343769</td>\n",
       "      <td>0.563935</td>\n",
       "      <td>16</td>\n",
       "      <td>105371409</td>\n",
       "      <td>LOC105371409</td>\n",
       "      <td>uncharacterized LOC105371409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>499336992</td>\n",
       "      <td>0.547440</td>\n",
       "      <td>12</td>\n",
       "      <td>5426</td>\n",
       "      <td>POLE</td>\n",
       "      <td>polymerase (DNA directed), epsilon, catalytic ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>499350441</td>\n",
       "      <td>0.487532</td>\n",
       "      <td>20</td>\n",
       "      <td>140876</td>\n",
       "      <td>FAM65C</td>\n",
       "      <td>family with sequence similarity 65, member C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>499315843</td>\n",
       "      <td>0.452432</td>\n",
       "      <td>4</td>\n",
       "      <td>285489</td>\n",
       "      <td>DOK7</td>\n",
       "      <td>docking protein 7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>499317231</td>\n",
       "      <td>0.417990</td>\n",
       "      <td>4</td>\n",
       "      <td>4085</td>\n",
       "      <td>MAD2L1</td>\n",
       "      <td>MAD2 mitotic arrest deficient-like 1 (yeast)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>499343767</td>\n",
       "      <td>0.341436</td>\n",
       "      <td>16</td>\n",
       "      <td>101927793</td>\n",
       "      <td>LOC101927793</td>\n",
       "      <td>uncharacterized LOC101927793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>499323096</td>\n",
       "      <td>0.302684</td>\n",
       "      <td>7</td>\n",
       "      <td>79778</td>\n",
       "      <td>MICALL2</td>\n",
       "      <td>MICAL-like 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>499329195</td>\n",
       "      <td>0.240802</td>\n",
       "      <td>9</td>\n",
       "      <td>85301</td>\n",
       "      <td>COL27A1</td>\n",
       "      <td>collagen, type XXVII, alpha 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>499334626</td>\n",
       "      <td>0.218078</td>\n",
       "      <td>12</td>\n",
       "      <td>6539</td>\n",
       "      <td>SLC6A12</td>\n",
       "      <td>solute carrier family 6 (neurotransmitter tran...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>499319747</td>\n",
       "      <td>0.082936</td>\n",
       "      <td>5</td>\n",
       "      <td>101926975</td>\n",
       "      <td>LOC101926975</td>\n",
       "      <td>uncharacterized LOC101926975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>499323166</td>\n",
       "      <td>0.079103</td>\n",
       "      <td>7</td>\n",
       "      <td>84629</td>\n",
       "      <td>TNRC18</td>\n",
       "      <td>trinucleotide repeat containing 18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>499330167</td>\n",
       "      <td>0.062469</td>\n",
       "      <td>10</td>\n",
       "      <td>2572</td>\n",
       "      <td>GAD2</td>\n",
       "      <td>glutamate decarboxylase 2 (pancreatic islets a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>499330118</td>\n",
       "      <td>0.057279</td>\n",
       "      <td>10</td>\n",
       "      <td>100130992</td>\n",
       "      <td>LOC100130992</td>\n",
       "      <td>uncharacterized LOC100130992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>499317260</td>\n",
       "      <td>0.018512</td>\n",
       "      <td>4</td>\n",
       "      <td>729338</td>\n",
       "      <td>CETN4P</td>\n",
       "      <td>centrin EF-hand protein 4, pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>499326780</td>\n",
       "      <td>-0.016994</td>\n",
       "      <td>8</td>\n",
       "      <td>51050</td>\n",
       "      <td>PI15</td>\n",
       "      <td>peptidase inhibitor 15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>499352783</td>\n",
       "      <td>-0.030682</td>\n",
       "      <td>X</td>\n",
       "      <td>5475</td>\n",
       "      <td>PPEF1</td>\n",
       "      <td>protein phosphatase, EF-hand calcium binding d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>499322427</td>\n",
       "      <td>-0.051990</td>\n",
       "      <td>6</td>\n",
       "      <td>9729</td>\n",
       "      <td>KIAA0408</td>\n",
       "      <td>KIAA0408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>499315162</td>\n",
       "      <td>-0.060485</td>\n",
       "      <td>3</td>\n",
       "      <td>100505385</td>\n",
       "      <td>IQCJ-SCHIP1</td>\n",
       "      <td>IQCJ-SCHIP1 readthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>499315537</td>\n",
       "      <td>-0.079266</td>\n",
       "      <td>3</td>\n",
       "      <td>6750</td>\n",
       "      <td>SST</td>\n",
       "      <td>somatostatin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>499342049</td>\n",
       "      <td>-0.175585</td>\n",
       "      <td>16</td>\n",
       "      <td>6650</td>\n",
       "      <td>CAPN15</td>\n",
       "      <td>calpain 15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>499309627</td>\n",
       "      <td>-0.175789</td>\n",
       "      <td>2</td>\n",
       "      <td>130502</td>\n",
       "      <td>TTC32</td>\n",
       "      <td>tetratricopeptide repeat domain 32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>499324505</td>\n",
       "      <td>-0.219573</td>\n",
       "      <td>7</td>\n",
       "      <td>6863</td>\n",
       "      <td>TAC1</td>\n",
       "      <td>tachykinin, precursor 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>499334295</td>\n",
       "      <td>-0.222514</td>\n",
       "      <td>11</td>\n",
       "      <td>283149</td>\n",
       "      <td>BCL9L</td>\n",
       "      <td>B-cell CLL/lymphoma 9-like</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>499348654</td>\n",
       "      <td>-0.428175</td>\n",
       "      <td>19</td>\n",
       "      <td>4059</td>\n",
       "      <td>BCAM</td>\n",
       "      <td>basal cell adhesion molecule (Lutheran blood g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>499336206</td>\n",
       "      <td>-0.644976</td>\n",
       "      <td>12</td>\n",
       "      <td>728084</td>\n",
       "      <td>LOC728084</td>\n",
       "      <td>uncharacterized LOC728084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>499335939</td>\n",
       "      <td>-0.670024</td>\n",
       "      <td>12</td>\n",
       "      <td>105369807</td>\n",
       "      <td>LOC105369807</td>\n",
       "      <td>uncharacterized LOC105369807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>499328351</td>\n",
       "      <td>-0.686543</td>\n",
       "      <td>9</td>\n",
       "      <td>728433</td>\n",
       "      <td>LOC728433</td>\n",
       "      <td>fibroblast growth factor 7 pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>499336276</td>\n",
       "      <td>-0.713233</td>\n",
       "      <td>12</td>\n",
       "      <td>642969</td>\n",
       "      <td>LOC642969</td>\n",
       "      <td>phosphoglycerate mutase 1 (brain) pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>499334792</td>\n",
       "      <td>-0.728323</td>\n",
       "      <td>12</td>\n",
       "      <td>1822</td>\n",
       "      <td>ATN1</td>\n",
       "      <td>atrophin 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>499328366</td>\n",
       "      <td>-1.312643</td>\n",
       "      <td>9</td>\n",
       "      <td>100113421</td>\n",
       "      <td>LOC100113421</td>\n",
       "      <td>fibroblast growth factor 7 pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>499347240</td>\n",
       "      <td>-1.326129</td>\n",
       "      <td>19</td>\n",
       "      <td>105372247</td>\n",
       "      <td>LOC105372247</td>\n",
       "      <td>uncharacterized LOC105372247</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id      coef chromosome  gene_entrez_id   gene_symbol   \n",
       "0   499348546  0.637636         19           23152           CIC  \\\n",
       "1   499343769  0.563935         16       105371409  LOC105371409   \n",
       "2   499336992  0.547440         12            5426          POLE   \n",
       "3   499350441  0.487532         20          140876        FAM65C   \n",
       "4   499315843  0.452432          4          285489          DOK7   \n",
       "5   499317231  0.417990          4            4085        MAD2L1   \n",
       "6   499343767  0.341436         16       101927793  LOC101927793   \n",
       "7   499323096  0.302684          7           79778       MICALL2   \n",
       "8   499329195  0.240802          9           85301       COL27A1   \n",
       "9   499334626  0.218078         12            6539       SLC6A12   \n",
       "10  499319747  0.082936          5       101926975  LOC101926975   \n",
       "11  499323166  0.079103          7           84629        TNRC18   \n",
       "12  499330167  0.062469         10            2572          GAD2   \n",
       "13  499330118  0.057279         10       100130992  LOC100130992   \n",
       "14  499317260  0.018512          4          729338        CETN4P   \n",
       "15  499326780 -0.016994          8           51050          PI15   \n",
       "16  499352783 -0.030682          X            5475         PPEF1   \n",
       "17  499322427 -0.051990          6            9729      KIAA0408   \n",
       "18  499315162 -0.060485          3       100505385   IQCJ-SCHIP1   \n",
       "19  499315537 -0.079266          3            6750           SST   \n",
       "20  499342049 -0.175585         16            6650        CAPN15   \n",
       "21  499309627 -0.175789          2          130502         TTC32   \n",
       "22  499324505 -0.219573          7            6863          TAC1   \n",
       "23  499334295 -0.222514         11          283149         BCL9L   \n",
       "24  499348654 -0.428175         19            4059          BCAM   \n",
       "25  499336206 -0.644976         12          728084     LOC728084   \n",
       "26  499335939 -0.670024         12       105369807  LOC105369807   \n",
       "27  499328351 -0.686543          9          728433     LOC728433   \n",
       "28  499336276 -0.713233         12          642969     LOC642969   \n",
       "29  499334792 -0.728323         12            1822          ATN1   \n",
       "30  499328366 -1.312643          9       100113421  LOC100113421   \n",
       "31  499347240 -1.326129         19       105372247  LOC105372247   \n",
       "\n",
       "                                            gene_name  \n",
       "0                   capicua transcriptional repressor  \n",
       "1                        uncharacterized LOC105371409  \n",
       "2   polymerase (DNA directed), epsilon, catalytic ...  \n",
       "3        family with sequence similarity 65, member C  \n",
       "4                                   docking protein 7  \n",
       "5        MAD2 mitotic arrest deficient-like 1 (yeast)  \n",
       "6                        uncharacterized LOC101927793  \n",
       "7                                        MICAL-like 2  \n",
       "8                       collagen, type XXVII, alpha 1  \n",
       "9   solute carrier family 6 (neurotransmitter tran...  \n",
       "10                       uncharacterized LOC101926975  \n",
       "11                 trinucleotide repeat containing 18  \n",
       "12  glutamate decarboxylase 2 (pancreatic islets a...  \n",
       "13                       uncharacterized LOC100130992  \n",
       "14              centrin EF-hand protein 4, pseudogene  \n",
       "15                             peptidase inhibitor 15  \n",
       "16  protein phosphatase, EF-hand calcium binding d...  \n",
       "17                                           KIAA0408  \n",
       "18                            IQCJ-SCHIP1 readthrough  \n",
       "19                                       somatostatin  \n",
       "20                                         calpain 15  \n",
       "21                 tetratricopeptide repeat domain 32  \n",
       "22                            tachykinin, precursor 1  \n",
       "23                         B-cell CLL/lymphoma 9-like  \n",
       "24  basal cell adhesion molecule (Lutheran blood g...  \n",
       "25                          uncharacterized LOC728084  \n",
       "26                       uncharacterized LOC105369807  \n",
       "27              fibroblast growth factor 7 pseudogene  \n",
       "28       phosphoglycerate mutase 1 (brain) pseudogene  \n",
       "29                                         atrophin 1  \n",
       "30              fibroblast growth factor 7 pseudogene  \n",
       "31                       uncharacterized LOC105372247  "
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_ids_ranking = (\n",
    "pd.DataFrame.from_dict({k:v for k,v in zip(support_ids, log_reg_coefs)}, orient = 'index', columns= ['coef'])\n",
    ".rename_axis('gene_id')\n",
    ".reset_index()\n",
    ".sort_values(by='coef', ascending= False)\n",
    ")\n",
    "gene_ids_ranking = gene_ids_ranking.merge(gene_data, on='gene_id')\n",
    "gene_ids_ranking"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quick look at DT feature performance\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "       0.00679705, 0.        , 0.10974137, 0.01604549, 0.        ,\n",
       "       0.        , 0.01944908, 0.01011498, 0.26058385, 0.        ,\n",
       "       0.03554424, 0.02726113, 0.        , 0.04936825, 0.03697756,\n",
       "       0.02106588, 0.        , 0.        , 0.03409347, 0.02852532,\n",
       "       0.        , 0.07655832, 0.10158606, 0.        , 0.        ,\n",
       "       0.16628796, 0.        ])"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Quick look at DT feature performance\n",
    "saved_models['DT'].feature_importances_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[False False False False False  True False  True False False False  True\n",
      "  True  True  True  True False False  True  True  True  True False  True\n",
      " False False  True  True False  True  True False]\n",
      "[17 14 12  9 13  1  6  1  8  5  3  1  1  1  1  1 16 11  1  1  1  1  2  1\n",
      " 10  4  1  1  7  1  1 15]\n"
     ]
    }
   ],
   "source": [
    "# RFE with decision tree\n",
    "estimator = models[1]\n",
    "selector = RFE(estimator)\n",
    "selector = selector.fit(X_train_, y_train)\n",
    "print(selector.support_)\n",
    "print(selector.ranking_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Top genes\n",
    "# support_ids = X_train.columns[selector.support_].to_list()\n",
    "support_ranking = saved_models['DT'].feature_importances_[selector.support_]\n",
    "\n",
    "gene_ids_ranking = (\n",
    "pd.DataFrame.from_dict({k:v for k,v in zip(support_ids, support_ranking)}, orient = 'index', columns= ['DT_ranking'])\n",
    ".rename_axis('gene_id')\n",
    ".reset_index()\n",
    ".sort_values(by='DT_ranking', ascending= False)\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>DT_ranking</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_entrez_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>499317231</td>\n",
       "      <td>0.260584</td>\n",
       "      <td>4</td>\n",
       "      <td>4085</td>\n",
       "      <td>MAD2L1</td>\n",
       "      <td>MAD2 mitotic arrest deficient-like 1 (yeast)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>499330118</td>\n",
       "      <td>0.166288</td>\n",
       "      <td>10</td>\n",
       "      <td>100130992</td>\n",
       "      <td>LOC100130992</td>\n",
       "      <td>uncharacterized LOC100130992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>499315162</td>\n",
       "      <td>0.109741</td>\n",
       "      <td>3</td>\n",
       "      <td>100505385</td>\n",
       "      <td>IQCJ-SCHIP1</td>\n",
       "      <td>IQCJ-SCHIP1 readthrough</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>499328366</td>\n",
       "      <td>0.101586</td>\n",
       "      <td>9</td>\n",
       "      <td>100113421</td>\n",
       "      <td>LOC100113421</td>\n",
       "      <td>fibroblast growth factor 7 pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>499328351</td>\n",
       "      <td>0.076558</td>\n",
       "      <td>9</td>\n",
       "      <td>728433</td>\n",
       "      <td>LOC728433</td>\n",
       "      <td>fibroblast growth factor 7 pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>499322427</td>\n",
       "      <td>0.049368</td>\n",
       "      <td>6</td>\n",
       "      <td>9729</td>\n",
       "      <td>KIAA0408</td>\n",
       "      <td>KIAA0408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>499323096</td>\n",
       "      <td>0.036978</td>\n",
       "      <td>7</td>\n",
       "      <td>79778</td>\n",
       "      <td>MICALL2</td>\n",
       "      <td>MICAL-like 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>499319747</td>\n",
       "      <td>0.035544</td>\n",
       "      <td>5</td>\n",
       "      <td>101926975</td>\n",
       "      <td>LOC101926975</td>\n",
       "      <td>uncharacterized LOC101926975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>499326780</td>\n",
       "      <td>0.034093</td>\n",
       "      <td>8</td>\n",
       "      <td>51050</td>\n",
       "      <td>PI15</td>\n",
       "      <td>peptidase inhibitor 15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>499323166</td>\n",
       "      <td>0.021066</td>\n",
       "      <td>7</td>\n",
       "      <td>84629</td>\n",
       "      <td>TNRC18</td>\n",
       "      <td>trinucleotide repeat containing 18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>499315537</td>\n",
       "      <td>0.019449</td>\n",
       "      <td>3</td>\n",
       "      <td>6750</td>\n",
       "      <td>SST</td>\n",
       "      <td>somatostatin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>499315843</td>\n",
       "      <td>0.010115</td>\n",
       "      <td>4</td>\n",
       "      <td>285489</td>\n",
       "      <td>DOK7</td>\n",
       "      <td>docking protein 7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>499309627</td>\n",
       "      <td>0.006797</td>\n",
       "      <td>2</td>\n",
       "      <td>130502</td>\n",
       "      <td>TTC32</td>\n",
       "      <td>tetratricopeptide repeat domain 32</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id  DT_ranking chromosome  gene_entrez_id   gene_symbol   \n",
       "0   499317231    0.260584          4            4085        MAD2L1  \\\n",
       "1   499330118    0.166288         10       100130992  LOC100130992   \n",
       "2   499315162    0.109741          3       100505385   IQCJ-SCHIP1   \n",
       "3   499328366    0.101586          9       100113421  LOC100113421   \n",
       "4   499328351    0.076558          9          728433     LOC728433   \n",
       "5   499322427    0.049368          6            9729      KIAA0408   \n",
       "6   499323096    0.036978          7           79778       MICALL2   \n",
       "7   499319747    0.035544          5       101926975  LOC101926975   \n",
       "8   499326780    0.034093          8           51050          PI15   \n",
       "9   499323166    0.021066          7           84629        TNRC18   \n",
       "10  499315537    0.019449          3            6750           SST   \n",
       "11  499315843    0.010115          4          285489          DOK7   \n",
       "12  499309627    0.006797          2          130502         TTC32   \n",
       "\n",
       "                                       gene_name  \n",
       "0   MAD2 mitotic arrest deficient-like 1 (yeast)  \n",
       "1                   uncharacterized LOC100130992  \n",
       "2                        IQCJ-SCHIP1 readthrough  \n",
       "3          fibroblast growth factor 7 pseudogene  \n",
       "4          fibroblast growth factor 7 pseudogene  \n",
       "5                                       KIAA0408  \n",
       "6                                   MICAL-like 2  \n",
       "7                   uncharacterized LOC101926975  \n",
       "8                         peptidase inhibitor 15  \n",
       "9             trinucleotide repeat containing 18  \n",
       "10                                  somatostatin  \n",
       "11                             docking protein 7  \n",
       "12            tetratricopeptide repeat domain 32  "
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_data_with_ranking = gene_ids_ranking.merge(gene_data, on='gene_id')\n",
    "gene_data_with_ranking = gene_data_with_ranking[gene_data_with_ranking['DT_ranking'] > 0]\n",
    "gene_data_with_ranking"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mads",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
